org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/util/NodeModelUtils.java

```diff
@@ -111,7 +112,14 @@ public static ICompositeNode getNode(/* @Nullable */ EObject object) {
 		if (object instanceof INodeReference) {
 			return ((INodeReference) object).getNode();
 		}
-		return (ICompositeNode) EcoreUtil.getExistingAdapter(object, ICompositeNode.class);
+		Adapter adapter = EcoreUtil.getExistingAdapter(object, ICompositeNode.class);
+		if (adapter instanceof ICompositeNode) {
+			return (ICompositeNode) adapter;
+		}
+		if (adapter instanceof INodeReference) {
+			return ((INodeReference) adapter).getNode();
+		}
```

Comment:
Maybe I don't understand the details, but wouldn't this be equivalent to

```java
Adapter adapter = EcoreUtil.getExistingAdapter(object, ICompositeNode.class);
if (adapter instanceof INodeReference) {
	return ((INodeReference) adapter).getNode();
}
return (ICompositeNode) adapter;
```

Reply:
I wanted to have the instanceof check for the ICompositeNode first since it's far more likely to be true. Doesn't matter in practice so I'll apply your suggestion.